packages/flutter/lib/src/cupertino/segmented_control.dart

```diff
@@ -169,14 +222,70 @@ class _SegmentedControlState<T> extends State<SegmentedControl<T>> {
   }
 
   void _onTap(T currentKey) {
-    if (currentKey != widget.groupValue) {
+    if (currentKey != widget.groupValue && currentKey == _pressedKey) {
       widget.onValueChanged(currentKey);
+      _pressedKey = null;
+    }
+  }
+
+  Color getTextColor(int index, T currentKey) {
+    if (_selectionControllers[index].isAnimating) {
+      return textColorTween.evaluate(_selectionControllers[index]);
+    } else if (widget.groupValue == currentKey) {
+      return CupertinoColors.white;
+    }
+    return CupertinoColors.activeBlue;
+  }
+
+  Color getBackgroundColor(int index, T currentKey) {
+    if (_selectionControllers[index].isAnimating) {
+      return _childTweens[index].evaluate(_selectionControllers[index]);
+    } else if (widget.groupValue == currentKey) {
+      return CupertinoColors.activeBlue;
+    } else if (_pressedKey == currentKey) {
+      return _kPressedBackground;
+    }
+    return CupertinoColors.white;
+  }
```

Comment:
generally speaking when you have if statements that end in returns, it's considered best practice to not have an "else", because "else" sort of implies that both sides will exit out of the other side. So the usual way to write these two functions would be something like:
```
  if (...)
    return ...;
  if (...)
    return ...;
  if (...)
    return ...;
  return ...;
```